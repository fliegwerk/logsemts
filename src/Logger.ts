import randomColor from 'randomcolor';
import Style from './types/Style';
import ComponentLogger from './ComponentLogger';
import { Options } from './types/Options';
import BrowserLogger from './BrowserLogger';

const componentBackgroundColor = 'lightgray';

export default class Logger {
	private availableLoggers: { [key: string]: ComponentLogger } = {};

	constructor(private options: Options) {}

	getSubsystemLogger(componentName: string) {
		// build new subsystem logger if no one is available
		if (!this.availableLoggers[componentName]) {
			this.availableLoggers[componentName] = new ComponentLogger(
				(type, style, ...args) => {
					const componentStyle = new Style(
						randomColor({ luminosity: 'dark', seed: componentName }),
						componentBackgroundColor
					);

					for (let fn of this.options.loggers)
						fn(type, style, componentName, componentStyle, ...args);
				}
			);
			this.getSubsystemLogger('Logger').debug('Created logger', componentName);
		}

		return this.availableLoggers[componentName];
	}
}

export { BrowserLogger };
