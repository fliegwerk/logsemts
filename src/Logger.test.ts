import Logger from './Logger';

describe('Logger', () => {
	beforeEach(() => {
		jest.mock('randomcolor');
	});

	it('should log to the console', () => {
		const logFn = jest.fn();

		const logger = new Logger({
			loggers: [logFn]
		});
		logger.getComponentLogger('Test Component').debug('Hello world');

		expect(logFn).toHaveBeenCalled();
		expect(logFn).toMatchSnapshot();
	});
});
