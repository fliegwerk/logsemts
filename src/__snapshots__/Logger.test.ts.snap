// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should log to the console 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "DEBUG",
      Object {
        "background": "black",
        "color": "lightgray",
        "css": "color: lightgray; background-color: black; font-weight: bold;",
      },
      "Logger",
      Object {
        "background": "lightgray",
        "color": "#a00",
        "css": "color: #a00; background-color: lightgray; font-weight: bold;",
      },
      "Created logger",
      "Logger",
    ],
    Array [
      "DEBUG",
      Object {
        "background": "black",
        "color": "lightgray",
        "css": "color: lightgray; background-color: black; font-weight: bold;",
      },
      "Logger",
      Object {
        "background": "lightgray",
        "color": "#a00",
        "css": "color: #a00; background-color: lightgray; font-weight: bold;",
      },
      "Created logger",
      "Test Component",
    ],
    Array [
      "DEBUG",
      Object {
        "background": "black",
        "color": "lightgray",
        "css": "color: lightgray; background-color: black; font-weight: bold;",
      },
      "Test Component",
      Object {
        "background": "lightgray",
        "color": "#a00",
        "css": "color: #a00; background-color: lightgray; font-weight: bold;",
      },
      "Hello world",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
